Modules/hmacmodule.c

```diff
@@ -0,0 +1,1713 @@
+    DECL_HACL_HMAC_HASH_KIND(sha3_256, SHA3_256)
+    DECL_HACL_HMAC_HASH_KIND(sha3_384, SHA3_384)
+    DECL_HACL_HMAC_HASH_KIND(sha3_512, SHA3_512)
+    /* Blake family */
+    DECL_HACL_HMAC_HASH_KIND(blake2s_32, Blake2S_32)
+    DECL_HACL_HMAC_HASH_KIND(blake2b_32, Blake2B_32)
+    /* Blake runtime family (should not be used statically) */
+    DECL_HACL_HMAC_HASH_KIND(vectorized_blake2s_32, Blake2S_128)
+    DECL_HACL_HMAC_HASH_KIND(vectorized_blake2b_32, Blake2B_256)
+#undef DECL_HACL_HMAC_HASH_KIND
+} HMAC_Hash_Kind;
+
+typedef Hacl_Streaming_Types_error_code hacl_errno_t;
+
+/* Function pointer type for 1-shot HACL* HMAC functions. */
+typedef void
+(*HACL_HMAC_compute_func)(uint8_t *out,
+                          uint8_t *key, uint32_t keylen,
+                          uint8_t *msg, uint32_t msglen);
+/* Function pointer type for 1-shot HACL* HMAC CPython AC functions. */
+typedef PyObject *
+(*PyAC_HMAC_compute_func)(PyObject *module, PyObject *key, PyObject *msg);
+
+/*
+ * HACL* HMAC minimal interface.
+ */
+typedef struct py_hmac_hacl_api {
+    HACL_HMAC_compute_func compute;
+    PyAC_HMAC_compute_func compute_py;
+} py_hmac_hacl_api;
+
+#if PY_SSIZE_T_MAX > UINT32_MAX
+#define Py_HMAC_SSIZE_LARGER_THAN_UINT32
+#endif
+
+/*
+ * Assert that 'LEN' can be safely casted to uint32_t.
+ *
+ * The 'LEN' parameter should be convertible to Py_ssize_t.
+ */
+#ifdef Py_HMAC_SSIZE_LARGER_THAN_UINT32
+#define Py_CHECK_HACL_UINT32_T_LENGTH(LEN)                  \
+    do {                                                    \
+        assert((Py_ssize_t)(LEN) <= UINT32_MAX_AS_SSIZE_T); \
+    } while (0)
+#else
+#define Py_CHECK_HACL_UINT32_T_LENGTH(LEN)
+#endif
+
+/*
+ * Call the HACL* HMAC-HASH update function on the given data.
+ *
+ * The magnitude of 'LEN' is not checked and thus 'LEN' must be
+ * safely convertible to a uint32_t value.
+ */
+#define Py_HMAC_HACL_UPDATE_CALL(HACL_STATE, BUF, LEN)          \
+    Hacl_Streaming_HMAC_update(HACL_STATE, BUF, (uint32_t)(LEN))
+
+/*
+ * Call the HACL* HMAC-HASH update function on the given data.
+ *
+ * On DEBUG builds, the 'ERRACTION' statements are executed if
+ * the update() call returned a non-successful HACL* exit code.
+ *
+ * The buffer 'BUF' and its length 'LEN' are left untouched.
+ *
+ * The formal signature of this macro is:
+ *
+ *     (HACL_HMAC_state *, uint8_t *, uint32_t, PyObject *, (C statements))
+ */
+#ifndef NDEBUG
+#define Py_HMAC_HACL_UPDATE_ONCE(                                           \
+    HACL_STATE, BUF, LEN,                                                   \
+    ALGORITHM, ERRACTION                                                    \
+)                                                                           \
+    do {                                                                    \
+        Py_CHECK_HACL_UINT32_T_LENGTH(LEN);                                 \
+        hacl_errno_t code = Py_HMAC_HACL_UPDATE_CALL(HACL_STATE, BUF, LEN); \
```

Comment:
Strictly speaking, the only error code returned by the update function is `Hacl_Streaming_Types_MaximumLengthExceeded` but this should not happen in practice. I can keep this pedantic check for future possible error codes on debug builds, but I can also just assume that there will never be a new possible error code and simplify the macros.